geovin/includes/class-add-product-type.php

```diff
@@ -0,0 +1,190 @@
+<?php
+/**
+ * Geovin Variable Product Type
+ */
+namespace Geovin;
+
+if ( ! defined( 'ABSPATH' ) ) {
+    return;
+}
+
+class Add_Product_Type {
+
+    /**
+     * Build the instance
+     */
+    public function __construct() {
+        add_action( 'woocommerce_loaded', array( $this, 'load_product_type' ) );
+        add_filter( 'product_type_selector', array( $this, 'add_type' ) );
+        register_activation_hook( __FILE__, array( $this, 'install' ) );
+
+        add_filter('woocommerce_product_class', array( $this, 'geovin_variable_product_type_class'), 10, 2);
+
+        add_action('admin_footer', array($this,'geovin_variable_product_type_data_tabs'));
+
+
+        add_filter('woocommerce_product_data_tabs', array($this,'geovin_variable_product_data_tabs_for_product'), 10, 1);
+
+        add_filter('woocommerce_delete_variations_on_product_type_change', array($this,'do_not_remove_variations'), 10, 4);
+        add_filter( 'woocommerce_data_stores', array($this,'geovin_variable_data_store'), 10, 1 );
+        //add_action( 'woocommerce_geovin_add_to_cart', 'woocommerce_variable_add_to_cart' );
+
+        add_action( 'wp_ajax_woocommerce_get_variation_from_sku', array( $this, 'get_variation_from_sku' ) );
+        add_action( 'wp_ajax_nopriv_woocommerce_get_variation_from_sku', array( $this, 'get_variation_from_sku' ) );
+
+        // WC AJAX can be used for frontend ajax requests.
+        add_action( 'wc_ajax_get_variation_from_sku', array( $this, 'get_variation_from_sku' ) );
+
+        add_filter( 'woocommerce_add_to_cart_handler', array( $this, 'filter_add_to_cart_type' ), 10, 2 );
+
+        add_action( 'woocommerce_after_variations_table', array( $this, 'add_cart_image_input' ) );
+
+        add_filter( 'woocommerce_add_cart_item_data', array( $this, 'add_cart_image_data' ), 10, 4 );
+
+        add_filter( 'woocommerce_cart_item_thumbnail', array( $this, 'adjust_cart_thumb' ), 10, 3 );
+
+    }
+
+    public function adjust_cart_thumb( $image, $cart_item, $cart_item_key ) {
+        if ( isset( $cart_item['image_to_use'] ) && $cart_item['image_to_use'] !== '' ) {
+            $image = '<img width="300" height="300" src="' . $cart_item['image_to_use'] . '" class="attachment-woocommerce_thumbnail size-woocommerce_thumbnail" alt="" loading="lazy" />';
+        }
+        return $image;
+    }
+
+    public function add_cart_image_data( $cart_item_data, $product_id, $variation_id, $quantity ) {
+        $cart_item_data['image_to_use'] = $_REQUEST['cart_image_to_use'];
```

Comment:
When processing responses from the `$_REQUEST` object, data should be sanitized, for instance with `sanitize_text_field` or through the use of `filter_input`.

Also, it is helpful to use either `$_POST` or `$_GET`  (or `INPUT_GET` or `INPUT_POST`) rather than a full request object to prevent data from being spoofed.